File: app/src/main/java/com/android/mathias/velocity/ActivitySettings.java
package com.android.mathias.velocity;


import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.preference.RingtonePreference;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.MenuItem;

import java.util.ArrayList;
import java.util.List;

public class ActivitySettings extends AppCompatActivity {
    private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = (preference, value) -> {
        String stringValue = value.toString();
        if (preference instanceof ListPreference) {
            ListPreference listPreference = (ListPreference) preference;
            int index = listPreference.findIndexOfValue(stringValue);
            preference.setSummary(index >= 0 ? listPreference.getEntries()[index] : null);
        } else if (preference instanceof RingtonePreference) {
            if (TextUtils.isEmpty(stringValue)) {
                preference.setSummary(R.string.pref_sound_none);
            } else {
                Ringtone ringtone = RingtoneManager.getRingtone(preference.getContext(), Uri.parse(stringValue));
                if (ringtone == null) {
                    preference.setSummary(null);
                } else {
                    String name = ringtone.getTitle(preference.getContext());
                    preference.setSummary(name);
                }
            }
        } else {
            preference.setSummary(stringValue);
        }
        //bindPreferenceSummaryToValue(preference);
        return true;
    };

    private static void bindPreferenceSummaryToValue(Preference preference) {
        preference.setOnPreferenceChangeListener(sBindPreferenceSummaryToValueListener);
        sBindPreferenceSummaryToValueListener.onPreferenceChange(preference, PreferenceManager.getDefaultSharedPreferences(preference.getContext()).getString(preference.getKey(), ""));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        setSupportActionBar(findViewById(R.id.toolbar));
        if(getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        getFragmentManager().beginTransaction().replace(R.id.frame_preferences, new FragmentPreferences()).commit();
    }

    public static class FragmentPreferences extends PreferenceFragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            addPreferencesFromResource(R.xml.preferences);
            setHasOptionsMenu(true);

            final ListPreference routePref = (ListPreference) findPreference("default_route");
            List<Route> routes = DBManager.getRoutes(getContext(), null);
            List<String> routeNames = new ArrayList<>();
            for (Route r : routes) { routeNames.add(r.getName()); }
            routeNames.add("None");
            routePref.setEntries(routeNames.toArray(new CharSequence[routeNames.size()]));
            routePref.setEntryValues(routeNames.toArray(new CharSequence[routeNames.size()]));
            routePref.setDefaultValue("None");

            bindPreferenceSummaryToValue(routePref);
            bindPreferenceSummaryToValue(findPreference("display_name"));
            bindPreferenceSummaryToValue(findPreference("notifications_sound"));
            bindPreferenceSummaryToValue(findPreference("sync_frequency"));
        }

        @Override
        public boolean onOptionsItemSelected(MenuItem item) {
            int id = item.getItemId();
            if (id == android.R.id.home) {
                getActivity().finish();
                return true;
            }
            return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/com/android/mathias/velocity/ActivitySettings.kt
package com.android.mathias.velocity


import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.android.mathias.velocity.db.DBManager

class ActivitySettings : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        setSupportActionBar(findViewById(R.id.toolbar))
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportFragmentManager.beginTransaction()
                .replace(R.id.frame_preferences, FragmentPreferences())
                .commit()
    }

    class FragmentPreferences : PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.preferences, rootKey)
            val routePref = findPreference("default_route") as ListPreference
            val routes = DBManager.getRoutes(context!!, null)
            val routeNames = ArrayList<String>()
            for (r in routes) routeNames.add(r.name!!)
            routeNames.add("None")
            routePref.entries = routeNames.toTypedArray<CharSequence>()
            routePref.entryValues = routeNames.toTypedArray<CharSequence>()
            routePref.setDefaultValue("None")
            bindSummaryToValue(routePref)
        }

        override fun onOptionsItemSelected(item: MenuItem): Boolean {
            val id = item.itemId
            if (id == android.R.id.home) {
                activity!!.finish()
                return true
            }
            return super.onOptionsItemSelected(item)
        }
    }

    companion object {
        private val valueListener =
                Preference.OnPreferenceChangeListener { preference: Preference, value: Any ->
                    val stringValue = value.toString()
                    when (preference) {
                        is ListPreference -> {
                            val index = preference.findIndexOfValue(stringValue)
                            preference.setSummary(if (index >= 0) preference.entries[index] else null)
                        }
                        else -> preference.summary = stringValue
                    }
                    true
                }

        private fun bindSummaryToValue(preference: Preference) {
            preference.onPreferenceChangeListener = valueListener
            valueListener.onPreferenceChange(preference, PreferenceManager
                    .getDefaultSharedPreferences(preference.context)
                    .getString(preference.key, ""))
        }
    }
}
